### DISCUSSION TITLE
Increase script size limit

### DISCUSSION THREAD

**Discussion by Flyer133** [CrashOverride85/zc95#81] [General]
Hello CrashOverride85!
Do you know il it is possible to increase the script size limit?
Maybe it is possible to upload only 2 scritps but more large?
The cmd show this screen:

D:\hide\zc95-main5\zc95-main\remote_access>lua_upload.py --script ./lua/enema33.lua --ip 192.168.1.100 --index 1
Connecting
Connection opened
Uploading...
Got error message:
    Script too large
Websocket connection closed
Websocket connection closed
Failed on line 699

Your box is incredible!
regards!

**Comment by CrashOverride85** [comment]
Glad to hear you're having fun with it 🙂️
That's really interesting, the size limit per script is ~24k, which I didn't expect anyone to ever reach. I'll see if it can be increased a bit, possibly having one slot that is significantly larger than the rest.

**Reply by Flyer133** [reply]
Hello!
I tried to modify the file luascript.s, and recompile the firmware, but the error is here.
Script 1: 11x4096=45096 (44k)
Script 5: 1x4096 (4k)

**Reply by CrashOverride85** [reply]
I've not tested it, but I think to increase script1 from 24k to 48k, you would need to change it from:
```
lua_script1_start:
.balign 4096
.fill 24576, 1, 0x00 /* 24k */
```
to:

```
lua_script1_start:
.balign 4096
.fill 49152, 1, 0x00 /* 48k */
```

**Reply by Flyer133** [reply]
I tried that, but the message "script too large" remains. Maybe i have a mistake...
I will retry today!
Thanks for your interest!!

**Reply by Flyer133** [reply]
Hello! 
I tried a new compilation with you settings but the error remains.
Maybe the compilation is bad?
I will search another source file who may describe the buffer parameters.
[zc95.zip](https://github.com/CrashOverride85/zc95/files/13634784/zc95.zip)

**Reply by CrashOverride85** [reply]
I've tried that change, and your attached fw, and both mostly work for for me if uploading to slot 0 (`--index 0`).  A new problem is that the upload utility is now giving the error `Didn't get any message, expected Ack` at the end as it isn't waiting long enough for the script to be checked/saved, but it is still uploading ok.

**Reply by Flyer133** [reply]
Hello CashOverride85!
The same for me since few weeks, impossible to finish to upload a large script.
But the message "script too large " remains with the modified firmware.
Are you thinking that the solution is only to modify the .py upload scripts?
Regards!
JR

**Comment by CrashOverride85** [comment]
This is a modified copy of the main zc95 fw that should have a 48k (instead of 24k) Lua slot 0:
[zc95.zip](https://github.com/CrashOverride85/zc95/files/14453756/zc95.zip)

It's built from this branch:
https://github.com/CrashOverride85/zc95/tree/v1_7_48k_slot_0
Which also includes a modifed lua_upload python script with an increased timeout so it can cope with the larger script.


So you can upload to it by running:
`python3 lua_upload.py --script script.lua --ip 192.168.1.xxx--index 0`

**Comment by Flyer133** [comment]
Thanks a lot CrashOverride85!

**Comment by Flyer133** [comment]
Hello! 
I always have my problem to upload large scripts. ;-(

When the code reach 24k, the box freeze after upload, but the script is uploaded.
When I continue, the box freeze, and nothing appens.

Any idea?

[NewProg3.4.txt](https://github.com/CrashOverride85/zc95/files/15000336/NewProg3.4.txt)
[logConsole.txt](https://github.com/CrashOverride85/zc95/files/15000337/logConsole.txt)

**Reply by CrashOverride85** [reply]
Hi thanks for reporting this, I'll be honest I hadn't tested it with scripts quite that complex!
It looks like the issue is a stack overflow when it tries to load the script after transferring it, probably due to the number of variables in the script. I'll see if I can increase the stack size at the weekend

**Reply by Flyer133** [reply]
Hello!
Thanks for your reponse.
I'm sure you're right.
When I copy a lot of printf("AZERTYAZERTYAZERTYAZERTYAZERTYAZERTYAZERTY"), all is OK.

I tried to compile the last firmware with:
limits.h
MAXSTACK= 400
but I have an error when I use cmake. ;-(

Regards 
JR

**Reply by Flyer133** [reply]
![image](https://github.com/CrashOverride85/zc95/assets/130572800/95fab6d1-9120-4ac3-aa40-04264d330d72)

**Reply by CrashOverride85** [reply]
Hmm, it builds ok for me, but I've had problems with PRId64 (and similar) before, if you change the line to:
`            printf("Loop time: %llu, batt: %d\n", timenow - loop_start, batt_percentage);`
it should work as before. I'll make this change too, as that format seems far better supported.

I don't think changing the Lua MAXSTACK will fix it, as I think the issue is the main C/Pico stack limit which is set by `PICO_STACK_SIZE` in `CMakeLists.txt`. 
Also I think it's already set to the maximum, and increasing it further needs changes else where.

If you disable the stack guards (`PICO_USE_STACK_GUARDS`) it does upload without crashing, which is why I'm fairly sure it is a stack overflow. But obviously removing that protection isn't a solution as it means something will be getting overwritten in memory (or is very close to being overwritten), which is _very_ likely to lead to stability problems and weird behaviour, if it doesn't outright crash with some scripts.
